File: packages/typescript/core/src/unicode.ts
import type { InsertionReason } from "./types.js";
import { graphemeSegments } from "unicode-segmenter/grapheme";
import {
  HAN_RANGES,
  LATIN_RANGES,
  MARK_RANGES,
  WHITE_SPACE_RANGES,
  containsCodePoint,
} from "./unicode-17.generated.js";

export const UNICODE_VERSION = "17.0.0";

export type BoundaryCategory = "Han" | "Latin" | "AsciiDigit" | "Whitespace" | "Other";

export interface Grapheme {
  readonly text: string;
  readonly start: number;
  readonly end: number;
  readonly category: BoundaryCategory;
}

export function classifyGrapheme(text: string): BoundaryCategory {
  const codePoints = [...text].map((scalar) => scalar.codePointAt(0));
  if (codePoints.some((codePoint) =>
    codePoint !== undefined && containsCodePoint(WHITE_SPACE_RANGES, codePoint))) {
    return "Whitespace";
  }
  if (
    codePoints.length === 1 &&
    codePoints[0] !== undefined &&
    codePoints[0] >= 0x30 &&
    codePoints[0] <= 0x39
  ) {
    return "AsciiDigit";
  }

  const base = codePoints.find((codePoint) =>
    codePoint !== undefined && !containsCodePoint(MARK_RANGES, codePoint));
  if (base === undefined) return "Other";
  if (containsCodePoint(HAN_RANGES, base)) return "Han";
  if (containsCodePoint(LATIN_RANGES, base)) return "Latin";
  return "Other";
}

export function segmentText(text: string): readonly Grapheme[] {
  return [...graphemeSegments(text)].map((part) => ({
    text: part.segment,
    start: part.index,
    end: part.index + part.segment.length,
    category: classifyGrapheme(part.segment),
  }));
}

export function insertionReason(
  left: BoundaryCategory,
  right: BoundaryCategory,
): InsertionReason | null {
  if ((left === "Han" && right === "Latin") || (left === "Latin" && right === "Han")) {
    return "hanLatin";
  }
  if (
    (left === "Han" && right === "AsciiDigit") ||
    (left === "AsciiDigit" && right === "Han")
  ) {
    return "hanAsciiDigit";
  }
  return null;
}

File: packages/swift/Sources/NaturalSpacingCore/Unicode.swift
enum BoundaryCategory: Sendable {
    case han
    case latin
    case asciiDigit
    case whitespace
    case other
}

struct Grapheme: Sendable {
    let text: String
    let start: Int
    let end: Int
    let category: BoundaryCategory
}

extension NaturalSpacing {
    static func segment(_ text: String) -> [Grapheme] {
        let boundaries = Grapheme17.boundaries(text)
        let utf16 = text.utf16
        return zip(boundaries, boundaries.dropFirst()).map { start, end in
            let lowerUTF16 = utf16.index(utf16.startIndex, offsetBy: start)
            let upperUTF16 = utf16.index(utf16.startIndex, offsetBy: end)
            let lower = String.Index(lowerUTF16, within: text)!
            let upper = String.Index(upperUTF16, within: text)!
            let value = String(text[lower..<upper])
            return Grapheme(
                text: value,
                start: start,
                end: end,
                category: classify(value)
            )
        }
    }

    static func insertionReason(
        left: BoundaryCategory,
        right: BoundaryCategory
    ) -> InsertionReason? {
        if (left == .han && right == .latin) || (left == .latin && right == .han) {
            return .hanLatin
        }
        if (left == .han && right == .asciiDigit)
            || (left == .asciiDigit && right == .han)
        {
            return .hanAsciiDigit
        }
        return nil
    }

    private static func classify(_ grapheme: String) -> BoundaryCategory {
        let scalars = Array(grapheme.unicodeScalars)
        if scalars.contains(where: { Unicode17.contains(Unicode17.whiteSpaceRanges, $0.value) }) {
            return .whitespace
        }
        if scalars.count == 1, let scalar = scalars.first, (48...57).contains(scalar.value) {
            return .asciiDigit
        }
        guard let base = scalars.first(where: {
            !Unicode17.contains(Unicode17.markRanges, $0.value)
        }) else {
            return .other
        }
        if Unicode17.contains(Unicode17.hanRanges, base.value) {
            return .han
        }
        if Unicode17.contains(Unicode17.latinRanges, base.value) {
            return .latin
        }
        return .other
    }
}
